components/RuiExpandCollapse/src/RuiExpandCollapse.ts

```diff
@@ -0,0 +1,194 @@
+   * style
+   */
+  private _triggerExpandAnimation(): void {
+    if (this._collapseableEl) {
+      var sectionHeight = this._collapseableEl.scrollHeight;
+      this._collapseableEl.style.height = sectionHeight + 'px';
+
+      const transitionEndHandler = (): void => {
+        if (this._collapseableEl) {
+          this._collapseableEl.style.height = '';
+          this._collapseableEl.removeEventListener('transitionend', transitionEndHandler);
+        }
+      }
+  
+      this._collapseableEl.addEventListener('transitionend', transitionEndHandler);
+    }
+  }
+
+  /**
+   * Initialises the expand collapse logic and styling,
+   * once complete makes the expand collapse visible
+   */
+  private _initialiseExpandCollapse(): void {
+    if (this.shadowRoot) {
+      this._collapseableEl = this.shadowRoot.querySelector('.details-container');
+
+      if (this._collapseableEl) {
+        
+        // need to set height initially if closed without triggering animation
+        if (!this.open) {
+          this._collapseableEl.style.height = '0px';
+        }
+        
+        let expandCollapse: HTMLElement | null = this.shadowRoot.querySelector('.expand-collapse');
+        if (expandCollapse) {
+          expandCollapse.style.opacity = '1';
+        }
+      }
+    }
+  }
+  
+  /**
+   * After initial render initialise expand collapse logic. Because
+   * we are animating slotted content, we have to wait for the 
+   * slot to be mounted
+   */
+  public firstUpdated(): void {
+    if (this.shadowRoot) {
+      this._detailsSlotEl = this.shadowRoot.querySelector('#details-slot');
+
+      if (this._detailsSlotEl) {
+        // when the slotted content changes we initialise expand collapse
+        // we need to wait for this because the animation of heigh calc 
+        // will only work once the slot and it's content have mounted and rendered
+        this._detailsSlotEl.addEventListener('slotchange', (): void => {
+          this._initialiseExpandCollapse();
+         });
+      }
+    }
+  }
+
+  public updated(changedProperties: Map<string, RuiExpandCollapsePropertyType>): void {
+    changedProperties.forEach((oldValue: RuiExpandCollapsePropertyType, propName: string): void => {
+      // detect change in open prop and trigger animation as necessary
+      if (propName === 'open' && this._collapseableEl) {
+        // transition from closed to open
+        if (this.open && !oldValue) {
+          this._triggerExpandAnimation();
+        }
+         
+        // transition from open to closed
+        if (!this.open && oldValue) {
+          this._triggerCollapseAnimation();
+        }
+      }
+    });
+  }
+
+  /**
+   * Render method
+   */  
+  public render(): TemplateResult {
+    return html`
+      <section class=${`expand-collapse${this.open ?  ' is-open' : '' }`}>
+        <div @click="${this._handleClick}" class="summary-container" aria-expanded=${`${this.open ? 'true': 'false'}`}>
```

Comment:
We should change this to a button so that it is keyboard accessible.

Here's a good example of an accessible accordion that might help: https://www.w3.org/TR/2019/NOTE-wai-aria-practices-1.1-20190207/examples/accordion/accordion.html

Reply:
@jennasalau It seems like this is just what we have to do, although the trade off is semantically we aren't really meant to have non-text elements within a button, so it would limit what content should be contained within the summary section.

Reply:
@jnelssonsmith You can have non-text elements (images) as the only element inside buttons, as long as you add `aria-label` to the button to describe the purpose.

Reply:
@ttaiyo I'm more concerned about the need to create custom layouts using divs within the button, but I guess with css grid now you could achieve a whole lot with just img, spans, text etc. Possibly I'm being over-dramatic, but I'm not really sure what kind of wacky expand collapse layouts users will be trying to conjure up

Reply:
@ttaiyo @jeffdowdle I have updated the code to more closely follow the aria example - The summary section is now a button and the details section is set to hidden when not expanded. Please review when possible :)

Reply:
@jnelssonsmith There's actually more to this than the `button`, the content region (in this case `.details-container`) needs to have `role=region` and `aria-labelledby` value matching ID of the button.

Check the full HTML example provided in the design pattern: https://www.w3.org/TR/2019/NOTE-wai-aria-practices-1.1-20190207/examples/accordion/accordion.html

Reply:
@ttaiyo aahhh I see, okay I've added those now